elide-async/src/main/java/com/yahoo/elide/async/service/AsyncQueryCancelThread.java

```diff
@@ -0,0 +1,126 @@
+
+    @Override
+    public void run() {
+        cancelAsyncQuery();
+    }
+
+    /**
+     * This method cancels queries based on threshold.
+     */
+    protected void cancelAsyncQuery() {
+
+        try {
+            TransactionRegistry transactionRegistry = elide.getTransactionRegistry();
+
+            Map<UUID, DataStoreTransaction> runningTransactionMap = transactionRegistry.getRunningTransactions();
+
+            //Running transaction UUIDs
+            Set<UUID> runningTransactionUUIDs = runningTransactionMap.keySet();
+
+            //Construct filter expression
+            PathElement statusPathElement = new PathElement(AsyncQuery.class, QueryStatus.class, "status");
+            List<QueryStatus> statusList = new ArrayList<QueryStatus>();
+            statusList.add(QueryStatus.CANCELLED);
+            statusList.add(QueryStatus.PROCESSING);
+            statusList.add(QueryStatus.QUEUED);
+            FilterExpression fltStatusExpression =
+                    new InPredicate(statusPathElement, statusList);
+
+            Collection<AsyncQuery> asyncQueryCollection =
+                    asyncQueryDao.loadAsyncQueryCollection(fltStatusExpression);
+
+            //Active AsyncQuery UUIDs
+            Set<UUID> asyncTransactionUUIDs = asyncQueryCollection.stream()
+                    .filter(query -> query.getStatus() == QueryStatus.CANCELLED
+                    || TimeUnit.SECONDS.convert(Math.abs(new Date(System.currentTimeMillis()).getTime()
+                            - query.getCreatedOn().getTime()), TimeUnit.MILLISECONDS) > maxRunTimeSeconds)
+                    .map(query -> UUID.fromString(query.getRequestId()))
+            .collect(Collectors.toSet());
+
+            //AsyncQuery UUIDs that have active transactions
+            Set<UUID> queryUUIDsToCancel = Sets.intersection(runningTransactionUUIDs, asyncTransactionUUIDs);
+
+            //AsyncQuery IDs that need to be cancelled
+            Set<String> queryIDsToCancel = queryUUIDsToCancel.stream()
+            .map(uuid -> {
+                return asyncQueryCollection.stream()
+                .filter(query -> query.getRequestId().equals(uuid.toString()))
+                .map(AsyncQuery::getId)
+                .findFirst().orElseThrow(IllegalStateException::new);
+            })
+            .collect(Collectors.toSet());
+
+            //Cancel Transactions
+            queryUUIDsToCancel.stream()
+               .forEach((uuid) -> {
+                   JsonApiDocument jsonApiDoc = new JsonApiDocument();
+                   MultivaluedMap<String, String> queryParams = new MultivaluedHashMap<String, String>();
+                   RequestScope scope = new RequestScope("", "query", NO_VERSION, jsonApiDoc,
+                           transactionRegistry.getRunningTransaction(uuid), null, queryParams,
+                           uuid, elide.getElideSettings());
+                   transactionRegistry.getRunningTransaction(uuid).cancel(scope);
```

Comment:
may be we should store transactionRegistry.getRunningTransaction(uuid) in a variable and check for null before calling cancel.